README.md
# DePART - Deep Learning for Predicting Anion Exchange Chromatography Retention Times.

DePART is a high-level python package for retention time prediction in hydrophilic strong anion exchange chromatography (hSAX). DePART uses several packages for the pre-processing of (MaxQuant) peptide tables and the final machine learning prediction (custom format peptide tables are also supported).There are convenience functions implemented to use the same preprocessing as used in the manuscript.
The core function of DePART is the training & prediction of the 'elution time' of peptides from hSAX prepfractionation. Elution time is in our workflow exchangeable with the fraction from an prefractionation experiment, e.g. a discrete number.



Getting Started
---

DePART requires the installation of the following packages and ideally python 3.5.

Dependencies
---
* numpy
* scipy
* matplotlib
* seaborn
* pyteomics
* sklearn
* pandas
* keras
* joblib
* Biopython

To install all dependencies use ``` pip install requirements.txt```

 Installation
 ----
 To install depart clone the repository and switch to a command line prompt.
 
```pip install depart```
 

 CLI interface
 ----
 DePART comes with a command line script which is located in the bin folder. Run the script via the command line and use 
 ```python bin/pyDePART.py --help``` to get all available options.
 
 Notes
 ----
 The current master & dev branch also support the prediction of RP-LC RT times.
 Changing the architecture and the parameters of neural network can be done
 by editing the RTMLlib. A few parameters can be changed by the function call,
 e.g. the activation function (swish).
 

Authors
----
* Sven Giese

License
----
This project is licensed under the MIT License - see the LICENSE.md file for details

CHANGES.txt
1.2.1, 04.09.2018 -- Changed Readme and package structure for installation.
1.2, 04.04.2018 -- Added slightly better API control for using CV in the depart executable.
1.1, 03.04.2018 -- Added full support for modifications and regression on RP-LC.
1.0, 15.01.2018 -- Initial release. Featuring the implementation used in the analytical chemistry paper.

DePART/learning/models.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 30 18:28:55 2018

@author: hanjo
"""

from keras import optimizers
from keras.layers import Dense, Dropout
from keras.models import Sequential


def SAX_Model(ini_mode="normal", optimizer="adam", loss="categorical_crossentropy", 
              act=["relu", "tanh", "relu"], dr1=0.0, dr2=0.0, dr3=0.0, 
              input_dim=218, output_dim=29):
    """
    Returns the best-performing neural network model from the manuscript.
    """
    model = Sequential()
    model.add(Dense(50, input_dim=input_dim, kernel_initializer=ini_mode, 
                    activation=act[0]))
    model.add(Dropout(dr1)) 
    model.add(Dense(40, kernel_initializer='normal', activation=act[1]))
    model.add(Dropout(dr2)) 
    model.add(Dense(35, kernel_initializer='normal', activation=act[2]))
    model.add(Dropout(dr3)) 
    model.add(Dense(output_dim, kernel_initializer='normal', activation='softmax'))
    # Compile model
    model.compile(loss=loss, optimizer=optimizer, 
                  metrics=['categorical_accuracy', 'accuracy'])
    return(model)


def FNN_Classifier(ini_mode="normal", optimizer="adam", 
                              loss="categorical_crossentropy", 
                              act=["relu", "tanh", "relu"],
                              dropout=[0.0, 0.0, 0.0], input_dim=218,
                              output_dim=29):
    """
    Returns the best-performing neural network model from the manuscript.
    """
    model = Sequential()
    model.add(Dense(50, input_dim=input_dim, kernel_initializer=ini_mode, 
                    activation=act[0]))
    model.add(Dropout(dropout[0])) 
    model.add(Dense(40, kernel_initializer='normal', activation=act[1]))
    model.add(Dropout(dropout[1])) 
    model.add(Dense(35, kernel_initializer='normal', activation=act[2]))
    model.add(Dropout(dropout[2])) 
    model.add(Dense(output_dim, kernel_initializer='normal', activation='softmax'))
    # Compile model
    model.compile(loss=loss, optimizer=optimizer, 
                  metrics=['categorical_accuracy', 'accuracy'])
    return(model)
    
    
def FNN_Regressor(ini_mode="normal", loss="mse", act=["relu", "tanh", "relu"],
                  dropout=[0.0, 0.0, 0.0], input_dim=218, output_dim=1,
                  lr=0.001, neurons=[50, 40, 35]):
    """
    Returns the best-performing neural network model from the manuscript.
    """
    model = Sequential()
    model.add(Dense(neurons[0], input_dim=input_dim, kernel_initializer=ini_mode, 
                    activation=act[0]))
    
    model.add(Dropout(dropout[0])) 
    model.add(Dense(neurons[1], kernel_initializer='normal', activation=act[1]))
    
    model.add(Dropout(dropout[1])) 
    model.add(Dense(neurons[2], kernel_initializer='normal', activation=act[2]))
    
    #output layer
    model.add(Dropout(dropout[2])) 
    model.add(Dense(1, kernel_initializer='normal', activation='relu'))
    # Compile model
    optimizer =  optimizers.Adam(lr=lr)
    model.compile(loss=loss, optimizer=optimizer, metrics=['mse'])
    #kernel_regularizer=regularizers.l2(0.01)
    return(model)


DePART/preprocessing/PrepFactory.py
# -*- coding: utf-8 -*-

import re
import sys

import numpy as np
import pandas as pd
from pyteomics import parser

from DePART.preprocessing import FeatureFactory as FF
from DePART.reader import ReaderFactory as RF


def to_minimal_df(df_in):
    """
    Extracts a minimal df partof the given dataframe to be used with the
    implemented machine learning workflow. Also ensures that the indices are 
    compareable
    """
    df1 = df_in.copy().reset_index()
    if "Modified sequence" in df_in.columns:
        df2 = df_in[["Sequence", "Fraction", "Modified sequence"]]
    else:
        df2 = df_in[["Sequence", "Fraction"]]
    return(df1, df2)
    
    
def preprocess_df(infile, reader="MaxQuant", min_obs=300):
    """
    Preprocess a MaxQuant dataframe. Removes duplicates by highest
    intensity, requires at least nobs per Fraction and returns
    a minimal df that is used for the machine learning part.
    
    Parameters:
    ------------------------
    df_evidence: df, Dataframe containing the columns Seuence and Fraction
    """
    
    #read data
    if reader == "MaxQuant":
        df_evidence = RF.MaxQuant_Reader(infile, remove_mods=False)
        
    elif reader =="CSV":
        df_evidence = RF.CSV_Reader(infile)
        
    else:
        sys.exit("Option {} is not supported. Please choose either \
                 MaxQuant or CSV.".format(reader))
    #process training data    
    #select most intense
    if "Intensity" in df_evidence.columns:
        df_filter = filter_low_int_ids(df_evidence)
    else:
        df_filter = df_evidence.copy()
    #minimum obs filter per fraction
    if min_obs > 0:
        df_filter = filter_obs(df_filter, min_obs=min_obs)    
    #get a minimal df for processing
    df_meta, df = to_minimal_df(df_filter)    
    #generate features
    Generator = FF.FeatureGenerator()
    df_features = Generator.create_all_features(df)
    return(df_meta, df, df_features)
    
    
def mark_most_intense_id(df_evidence):
    """
    Function to filter peptides according to their identified fractions.
    
    Pepties were filtered according to the following criteria.
    1) unique peptides were kept
    2) peptides that were identified in 2 fractions were kept if
        the fractions were adjacent and disregarded if they were not.
        In the first case the peptide with the highest intensity wins and will
        be added to the data.
    3) peptides that were identified in more than 2 fractions were disregarded
    """
    indic_dic = {}
    #clean evidence
    #Peptide ID is the right identifier for the groups of peptides
    indicator = [False] * df_evidence.shape[0]
    
    for ii, grpi in df_evidence.groupby("Peptide ID"):
        nfractions = grpi.shape[0]
        #keep peptide, only identified in one fraction
        if nfractions == 1:
            indicator[grpi.index[0]] = True
            #indic_dic[ii] = "Keep"
        
        #exactly two fractions
        elif nfractions == 2:
            
            #has no intensity? just keep first occurrence
            if grpi.Intensity.dropna().shape[0] == 0:
                keepid = grpi.index[0]
            else:
                #depends on intensity whcih peptide id to take
                nuniquefracs = len(np.unique(grpi.Fraction))
                #case one: same fraction
                if nuniquefracs == 1:
                    keepid = np.argmax(grpi.Intensity)
                #case two: different fraction
                else:
                    #adjacent fraction
                    if np.abs(np.diff(grpi.Fraction.sort_values())[0]) == 1:
                        keepid = np.argmax(grpi.Intensity)
                    #further away
                    else:
                        continue
                        #indic_dic[ii] = "Not Adjacent"
                #store results
                indic_dic[keepid] = "Keep"
                indicator[keepid] = True
                
        #ignore all peptides with ids in more than two fractions
        else:
            #indic_dic[ii] = ">2Fractions"
            pass
    return(indicator)
    
    
def filter_low_int_ids(df_in):
    """
    Description.
    
    Requires the input format to have the columns:
        Sequence, Fraction, Intensity, Peptide ID
    """
    req_cols = set(["Sequence", "Fraction", "Intensity", "Peptide ID"])
    if len(set(df_in.columns) & req_cols)==4:
        #filter only peptides that occur in 1 or 2 fractions.
        #for two fractions take the fraction with the highest intensity
        df_in["Fraction.Filter"] = mark_most_intense_id(df_in)
        return(df_in[df_in["Fraction.Filter"]==True])
    else:
        print ("Your data frame is missing one of the required columns:\
               Sequence, Fraction, Intensity, Peptide ID")
        return(None)


def filter_obs(df_in, min_obs=300):
    """
    Filters the data based on the minimum number of observations per fraction.
    
    Parameter:
    .-------------------------
    """
    df_model = df_in.copy()
    min_frac = df_model.Fraction.min()
    max_frac = df_model.Fraction.max() + 1
    print ("NRows BEFORE min obs filter: {}".format(df_model.shape[0]))
    obs = {i:j for i,j in zip(np.arange(min_frac, max_frac), 
                              np.bincount(df_model.Fraction)[1:])}
    df_model["nobs"] = [obs[i] for i in df_model.Fraction]
    df_model = df_model[df_model["nobs"] >=min_obs]
    print ("NRows AFTER min obs filter: {}".format(df_model.shape[0]))
    return(df_model)
    
def remove_brackets(seq):
    """
    Removes all brackets (and underscores...) from protein sequences.
    """
    return(re.sub("[\(\)\[\]_\-]", "", seq))
  
   
#%%
def extract_modifications(sequences, verbose=False):
    """
    Performs simple checks on the sequences. Assumes that the
    sequences has the "Amod" format, where A is an amino acid and
    mod a lower case modification. All brackets are removed.
    Mods need to be strictly lower case.
    
    sequences = ["ELVIS", "PEPT(phos)ID(bs3)E", 
                 "ELVISCcmPEPTIbs3DE", "PEPT[unimod:21]IDE"]
    """
    
    sequences = sequences.apply(FF.simply_alphabet)
    #init data structures
    patterns = []
    npatterns = 0
    all_patterns = {}
    detected_mods = []
    
    #some formatting    
    AA_set = set(parser.std_amino_acids)
    test_val = np.zeros(len(sequences))
    sequences = np.array(sequences)
    
    #matches all nterminal mods, e.g. glD or acA
    nterm_pattern = re.compile(r'^([a-z]+)([A-Z])', flags=re.MULTILINE)
    
    #test each sequence for non-AA letters
    for ii, seqi in enumerate(sequences):
        nterm_match = re.findall(nterm_pattern, seqi)
        #nterminal acetylation
        if len(nterm_match) == 0:
            pass
        else:
            #we can have multiple modifications on the sme residue in different peptides
            if nterm_match[0][1] in all_patterns:
                #only sotre if new modifications
                if nterm_match[0][0] in all_patterns[nterm_match[0][1]]:
                    pass
                else:
                    all_patterns[nterm_match[0][1]].append(nterm_match[0][0])
            else:
                all_patterns[nterm_match[0][1]] = [nterm_match[0][0]]
                
            seqi = seqi.replace(nterm_match[0][0], "")
        test_val[ii] = np.sum([1 for aai in seqi if aai not in AA_set])
    
    #keep the problematic ones
    prob_sequences = sequences[np.where(test_val>0)]
    #replace all kind of brackets
    prob_sequences = pd.Series([remove_brackets(seqi) for seqi in prob_sequences])
    
    if len(prob_sequences) == 0:
        print("There were no modifications found in your data.")
        return({}, None)

    for probi in prob_sequences:      
        #other mods
        pattern = re.compile("([A-Z])([^A-Z]+)")
        matches = re.findall(pattern, probi)
        if verbose:
            print (probi)
            print (matches)
            
        aa, mod = zip(*matches)
        patterns.append(mod)
        npatterns += len(mod)
        for aai,modi in zip(aa, mod):
            all_patterns[aai] = modi
            
        if len(mod) == 0:
            detected_mods.append(None)
        else:
            detected_mods.append(mod)
    
    if npatterns > 0:
        print ("We identified the following modifications: {}".format(all_patterns.items()))
    else:
        print ("There were no identifyable amino acids/modifications in your data: {}. We cannot deal with these...".format(",".join(prob_sequences.head().values)))   
    return(all_patterns, detected_mods)
   
def replace_numbers(seq):
    rep = {"1":"one",
          "2":"two",
          "3":"three",
          "4":"four",
          "5":"five",
          "6":"six",
          "7":"seven",
          "8":"eight",
          "9":"nine",
          "0":"zero"}
    pattern = re.compile("|".join(rep.keys()))
    return(pattern.sub(lambda m: rep[re.escape(m.group(0))], seq))
    
def remove_lower_letters(seq):
    """
    Removes lower capital letters from the sequence.
    """
    return(re.sub("[a-z]", "", seq))
    

def to_unmodified_sequence(seq):
    """
    Removes lower capital letters, brackets from the sequence.
    """
    return(re.sub("[^[A-Z]", "", seq))
    
def rewrite_modsequences(seq):
    """
    Rewrites modified sequences to modX format. requires the
    input to be preprocessed such that no brackets are in the sequences.
    
    Meant to be used via apply.
    seq = "ELVIS"
    seq = "ELVISCcmASD"
    """
    #TODO, replace numbers with text!  
    return(re.sub("([A-Z])([^A-Z]+)", r'\2\1', seq))
    
    
def replace_nterm_mod(seq):
    """
    Removes the nterminal modification.
    """
    return(re.sub(r'^([a-z]+)([A-Z])', r'\2', seq, flags=re.MULTILINE))
    
    

DePART/learning/processing.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 30 21:06:31 2018

@author: hanjo
"""
import numpy as np
import pandas as pd
import seaborn as sns
from joblib import Parallel, delayed
from keras import backend as K
from keras.utils import np_utils
from keras.utils.generic_utils import get_custom_objects
from pyteomics import achrom
from scipy import stats
from sklearn.metrics import mean_absolute_error, auc, mean_squared_error, r2_score, f1_score, accuracy_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

# cross imports
from DePART.learning import models


def swish(x):
    return (K.sigmoid(x) * x)

get_custom_objects().update({'swish': swish})


def eval_predictions_complex(y_test, y_pred, name, get_metrics=False):
    """
    Evaluation of the prediction. Generates the following metrics in a list:
        - mean absolute error
        - men squared error
        - f1
        - accuracy
        - mean distance (predicted vs. true)
        - std distance (""---------------")
        - eq_pred - predicted == true
        - oneoff -fiveoff - classification distance
        - correlation - pearsonr
        - name - method name (e.g. OLS, OLR, etc.)
        
    Parameters:
    ---------------------------------------
    y_test: ar,
            test observations
    y_pred: ar,
            predicted observations
    name: str, 
          method name
    get_metrics: bool,
                if True, only the column names for the metrics are rturned
                if False, computes the actual metrics
        
    """
    def help_diff(y_pred, y_test, t=0):
        """
        Create indicator to measure how far off the prediction is.
        """
        if t == 0:
            return(sum([1. if i==j else 0 for i,j in zip(y_test, [int(i) for i in np.round(y_pred)])])/y_test.shape[0] * 100)
        else:
            return(sum([1. if np.abs(i-j) <= t else 0 for i,j in zip(y_test, [int(i) for i in np.round(y_pred)])])/y_test.shape[0] * 100)
                
    if get_metrics:
        return(["MSA", "MSE", "R^2", "F1 (weighted)", "Accuracy", "mean. dist", 
                 "std. dist", "AUC",
                  "0-off-pred", "1-off-pred", "2-off-pred", "3-off-pred", 
                  "5-off-pred", "Correlation", "Method"])
    else:
        #compute difference and count occurrences
        diff = np.abs((y_test-y_pred).astype(int))
        mae = mean_absolute_error(y_test, y_pred)
        mse = mean_squared_error(y_test, y_pred)
        aucres = auc(y_test, y_pred, reorder=True)
        rsq = r2_score(y_test, y_pred)
        f1 = f1_score(np.round(y_test), [int(i) for i in np.round(y_pred)], average="weighted")
        acc = accuracy_score(np.round(y_test), np.round(y_pred))
        mean_dist =  np.mean(diff)
        std_dist = np.std(diff)
        eq_pred = help_diff(y_pred, y_test, t=0) 
        oneoff = help_diff(y_pred, y_test, t=1)
        twooff = help_diff(y_pred, y_test, t=2)
        throff = help_diff(y_pred, y_test, t=3)
        fiveoff = help_diff(y_pred, y_test, t=4)
        correlation = stats.pearsonr(y_test, y_pred)[0]
        return([mae, mse, rsq, f1, acc, mean_dist,std_dist, aucres, eq_pred,
                oneoff, twooff, throff, fiveoff, correlation, name])
    

def cv_splitter(split=5):
    """
    Prepares kfold cross-validation with reproduceable split.
    
    Parameters:
    --------------------------
            
    split: int,
            number of folds to return (iterator)
    """
    kf = KFold(n_splits=split, random_state=42)
    return(kf)


def fit_model(clf, traindf, testdf, trainy, testy, name, epochs=100,
              batch_size=512, return_pred=False, scale=False,
              regression=False, diagnostics=True):
    """
    Generic function that fits a model and retrieves the error rates.
    
    Parameters:
    --------------------------
    clf: classifier (sklearn),
         a classifier object that is either a sklearn classifier or a supported
         other object (e.g pyteomics, keras). This can only be
         distinguished if a corect name is used ("Pyteomics" or "Keras").
        
    traindf: ar-like,
            Can either be a numpy matrix or pandas dataframe with the
            training data.
    
    testdf: ar-like,
            Can either be a numpy matrix or pandas dataframe with the
            testing data.
            
    trainy: ar-like,
            Vector with y variables. Needs to be adapted for Keras!
            (softmax)

    testy: ar-like,
            Vector with y variables. Needs to be adapted for Keras!
            (softmax)
    name: str,
            Idstring for the classifier. The two hard coded clauses
            are "Pyteomics" which uses the pyteomics module to predict
            retention times and "Keras" which uses the keras
            framework for prediction. All other names don't have a special
            meaning. There, the clf object is more important and needs to be
            similar to the sklearn API.
    nomod: bool,
          if True, supports modifications in the pyteomics model.
          return_pred: bool,
          if True, returns the predicted target variables (default:False)     
 
    #debug:
    for train_index, test_index in kf.split(X_train_full):
        break
    clf = initialize_neural_network()
    traindf = np.array(X_train_full.iloc[train_index])
    testdf = np.array(X_train_full.iloc[test_index])
    trainy = np_utils.to_categorical(encoder.transform(y_train_full.iloc[train_index]))
    testy = np_utils.to_categorical(encoder.transform(y_train_full.iloc[test_index]))
    name = "Keras"
    
    """
    if scale:
        scaler = StandardScaler()
        scaler.fit(traindf)
        traindf = scaler.transform(traindf)
        testdf = scaler.transform(testdf)
        
    #train 
    clf.fit(traindf, trainy, epochs=epochs, batch_size=batch_size, verbose=0)
    
    #predict
    if not regression:
        yhat_train = clf.predict(traindf).argmax(axis=1)
        yhat_test = clf.predict(testdf).argmax(axis=1)
        
        trainy = trainy.argmax(axis=1)
        testy = testy.argmax(axis=1)        
    else:
        yhat_train = np.ravel(clf.predict(traindf))
        yhat_test = np.ravel(clf.predict(testdf))
        

    #evaluate
    res_train = eval_predictions_complex(trainy, yhat_train, name+"_Train")
    res_test = eval_predictions_complex(testy, yhat_test, name+"_Test")
    res_df = pd.DataFrame([res_train, res_test])  
    res_df.columns = eval_predictions_complex(None, None, None, True)
    
    if diagnostics:
        sns.jointplot(trainy, yhat_train)
        sns.jointplot(testy, yhat_test)
                
    if return_pred:
        return(clf, res_df, yhat_train, yhat_test)
    else:
        return(clf, res_df)
    
    
def process_df(ml_df):
    """
    Split df into datastructures that are useable with sklearn. So essentially,
    removing the sequence column and target column from the dataframe.
    
    Parameters:
    -------------------------------
    ml_df: ml_df,
            any dataframe where the Sequence and Fraction column should be
            removed.
    """
    tmp_df = ml_df.copy()
    seqs = tmp_df.Sequence
    fractions = tmp_df.Fraction
    
    if "PyteomicsSequence" in tmp_df.columns:
        del tmp_df["PyteomicsSequence"]
        
    if "Modified sequence" in tmp_df.columns:
        del tmp_df["PyteomicsSequence"]
        
    if "Sequence" in tmp_df.columns:
        del tmp_df["Sequence"]
        
    if "Fraction" in tmp_df.columns:
        del tmp_df["Fraction"]
        
    if "Score" in tmp_df.columns:
        del tmp_df["Score"]
        
    return(tmp_df, seqs, fractions)
    
        
def cross_validation(train_df, valid_df, name="", nkfold=5, n_jobs=5, nn_args={},
                     use_joblib=True, epochs=100, batch_size=512, verbose=False,
                     model=models.SAX_Model, loss="categorical_accuracy"):
    """
    Function for cross-validation on the training data and evaluation 
    on the hold-out set. It stores the predictions and computes more than
    the usual error metrics. For short evaluations use the build in keras
    cv function.
    
    Parameters:
    --------------------------
    train_df: df,
              training data
    valid_: df,
              validation data              
    nkfolds: int,
            number of folds
    n_jobs:int,
            number of jobs for parallel execution of cv
    nn_args: dict,
            keywrd arguments for the neural network init
    """
    def identity(x):
        return(x)
    
    print ("Training Size: {}".format(train_df.shape))

    #format the dtaframe for ML 
    X_train_full, Seqs_train_full, y_train_full = process_df(train_df)
    
    if loss == "categorical_accuracy":
        # encode class values as integers
        encoder = LabelEncoder()
        encoder.fit(y_train_full)
        encoded_y = pd.DataFrame(np_utils.to_categorical(encoder.transform(y_train_full)))
        encoded_y.index = y_train_full.index
        regression=False
        output_dims = len(np.unique(train_df.Fraction))
    else:
        #regression
        #identiy encoder
        encoded_y = y_train_full
        regression = True
        output_dims = 1
        
    #output dims depending on the number of fractions
    output_dims = len(np.unique(train_df.Fraction))
    input_dims = X_train_full.shape[1]
    
    #init cross-validation
    kf = cv_splitter(nkfold)
    results_store = []
    if use_joblib:
        results_store.extend(Parallel(n_jobs=n_jobs)(delayed(fit_model)
                        (model(output_dim=output_dims, input_dim=input_dims,
                               loss=loss, **nn_args), 
                         np.array(X_train_full.iloc[train_index]),
                         np.array(X_train_full.iloc[test_index]),
                         np.array(encoded_y.iloc[train_index]),
                         np.array(encoded_y.iloc[test_index]),
                         name,
                         regression=regression) for train_index, test_index in kf.split(X_train_full)))
    else:
        for train_index, test_index in kf.split(X_train_full):
            #prep data
            tmp_train = np.array(X_train_full.iloc[train_index])
            tmp_test = np.array(X_train_full.iloc[test_index])
            
            #ydata
            tmp_ytrain = np.array(encoded_y.iloc[train_index])
            tmp_ytest = np.array(encoded_y.iloc[test_index])
            
            tmp_model = model(output_dim=output_dims, input_dim=input_dims, 
                              loss=loss, *nn_args)
                       
            results_store.append(fit_model(tmp_model, tmp_train, tmp_test,
                                           tmp_ytrain, tmp_ytest, name,
                                           epochs=epochs, 
                                           regression=regression,
                                           batch_size=batch_size))
                         
                         
    classifier, res_ar = zip(*results_store)
    eval_df = pd.concat(res_ar)
    eval_df["Classifier"] = [i.split("_")[0] for i in eval_df["Method"]]
    eval_df = eval_df.groupby("Method").aggregate([np.mean, stats.sem])
    eval_df["data"] = [i.split("_")[1] for i in eval_df.index]
    eval_df["min_acc"] = eval_df["0-off-pred"]["mean"] - eval_df["0-off-pred"]["sem"]
    eval_df["max_acc"] = eval_df["0-off-pred"]["mean"] + eval_df["0-off-pred"]["sem"]
    eval_df["diff"] = eval_df["max_acc"] -  eval_df["min_acc"]
    eval_df = eval_df.round(3)    
    return(eval_df)
    
    
def train_validation(train_df, valid_df, epochs=100, batch_size=512, plot=False,
                     nn_args={}):
    """
    Wrapper for training on the complete training data and evaluating the
    performance on the hold-out set.
    
    Parameter:
    -------------------
    train_df: df,
                train df with features and
    valid_df: df,
               validation df with features
               
    Returns:
    -------------------
    res_df: metrics
    nnmodel: neural network model
    """
    #format the dtaframe for ML 
    X_train_full, Seqs_train_full, y_train_full = process_df(train_df)
    X_valid_full, Seqs_valid_full, y_valid_full = process_df(valid_df)    
    
    # encode class values as integers
    encoder = LabelEncoder()
    encoder.fit(y_train_full)

    #output dims depending on the number of fractions
    output_dims = len(np.unique(train_df.Fraction))
    input_dims = X_train_full.shape[1]
    
    nnmodel = models.SAX_Model(output_dim=output_dims, 
                               input_dim=input_dims, **nn_args)
    print (nnmodel.summary())
    history = nnmodel.fit(np.array(X_train_full),
                          np_utils.to_categorical(encoder.transform(y_train_full)),
                          epochs=epochs, batch_size=batch_size)
    
    #fit the model to the complete training data
    yhat_train_prob = nnmodel.predict(np.array(X_train_full))
    yhat_train_disc = yhat_train_prob.argmax(axis=1) + 1

    yhat_val_prob = nnmodel.predict(np.array(X_valid_full))
    yhat_val_disc = yhat_val_prob.argmax(axis=1) + 1
            
    #evaluate
    res_train = pd.DataFrame(eval_predictions_complex(y_train_full, yhat_train_disc, "keras_Train"))
    res_valid = pd.DataFrame(eval_predictions_complex(y_valid_full, yhat_val_disc, "keras_Valid"))
    
    res_df = pd.concat([res_train.transpose(), res_valid.transpose()])
    res_df.columns = eval_predictions_complex(None, None, None, True)
    
    if plot:
        x = np.arange(-4, 30, 1)
        ax1 = sns.jointplot(x=y_valid_full, y=yhat_val_disc, kind="kde",
                            xlim=(-4, 30 ), ylim=(-4, 30 ))
        ax1.set_axis_labels(xlabel="True Fraction", ylabel="Prediction")
        ax1.ax_joint.plot(x, x, '-k')
    print ("Results on the validation data:")
    print (res_df)
    return(res_df, nnmodel, history)
    

 
def pyteomcis_snippets(traindf, trainy, testdf, nomod=True):
    """
    """
    #pyteomics
    clf = achrom.get_RCs_vary_lcp([str(i).replace("U", "C") 
                                                for i in traindf], trainy)
    print ("Pyteomics LCP: {}".format(clf["lcp"]))
    yhat_train = [achrom.calculate_RT(i, clf, raise_no_mod=nomod) for i in 
                  traindf]
    yhat_test = [achrom.calculate_RT(i, clf, raise_no_mod=nomod) for i in 
                 testdf]
    return(yhat_train, yhat_test)
    
    
    